### PR TITLE
[Quant] Enable XNNPACK ops in QNNPACK BackendConfig

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85863

**Summary:** This commit enforces the following constraints on the
QNNPACK BackendConfig:

- `quant_min_lower_bound` = -127 for qint8 weight
- `quant_max_upper_bound` = 127 for qint8 weight
- `scale_min_lower_bound` = 2 ** -12 for qint8 activations and weight

These constraints will enable users to use this BackendConfig with
faster XNNPACK quantized ops. They are also consistent with the
existing settings in `default_symmetric_qnnpack_qconfig` and its
per_channel and QAT variants. For more detail on why these exact
values were chosen, please see the description of
https://github.com/pytorch/pytorch/pull/74396.

Note that there are currently no restrictions on the qscheme in
DTypeConfig. This should be added in the future to further enforce
the restriction that the weights must be quantized with either
per_tensor_symmetric or per_channel_symmetric.

Existing default QConfigs such as `get_default_qconfig("qnnpack")`
and `get_default_qat_qconfig("qnnpack")` will continue to be
supported, but only for the existing dtypes, e.g. quint8 activations
for weighted ops like linear and conv. In the future, we should
revisit whether to enable XNNPACK ops using these QConfigs as well.

**Test Plan:**

python test/test_quantization.py TestQuantizeFx.test_qnnpack_backend_config

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

### CODE REVIEW

File: torch/ao/quantization/backend_config/qnnpack.py [hunk 1]
```diff
@@ -11,63 +11,112 @@
...
+weight_qint8_with_constraints = _get_weight_dtype_with_constraints(torch.qint8)
+weight_quint8_with_constraints = _get_weight_dtype_with_constraints(torch.quint8)
+weight_quint4x2_with_constraints = _get_weight_dtype_with_constraints(torch.quint4x2)
+weight_fp16_with_constraints = _get_weight_dtype_with_constraints(torch.float16)
+
+# weighted op
+
+qnnpack_weighted_op_qint8_dtype_config = DTypeConfig(
+    input_dtype=act_qint8_with_constraints,
+    output_dtype=act_qint8_with_constraints,
```
Comment (andrewor14) [inline torch/ao/quantization/backend_config/qnnpack.py:45]: These qint8 `DTypeConfig`s are added to work with `default_symmetric_qnnpack_qconfig` and its per_channel and QAT equivalents

File: torch/ao/quantization/fx/utils.py [hunk 2]
```diff
@@ -972,11 +972,16 @@ def _activation_post_process_satisfies_dtype_config_constraints(
             activation_post_process: Union[ObserverBase, FakeQuantize],
             dtype_with_constraints: DTypeWithConstraints,
             debug_string: str) -> bool:
-        app_quant_min = getattr(activation_post_process, "quant_min", None)
-        app_quant_max = getattr(activation_post_process, "quant_max", None)
+        # For FakeQuantize, the attributes are set in the observer inside the FakeQuantize
+        if isinstance(activation_post_process, FakeQuantize):
+            observer = activation_post_process.activation_post_process
+        else:
+            observer = activation_post_process
```
Comment (andrewor14) [inline torch/ao/quantization/fx/utils.py:979]: This is a small bug fix for https://github.com/pytorch/pytorch/pull/85200, which did not handle `FakeQuantize` correctly

Comment (andrewor14) [inline torch/ao/quantization/fx/utils.py:979]: This is a bug fix for https://github.com/pytorch/pytorch/pull/85200, which did not handle `FakeQuantize` correctly

File: torch/ao/quantization/backend_config/qnnpack.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/backend_config/qnnpack.py:45]: I'm wondering if we want to add `qscheme` support to this DTypeConfig as well

File: torch/ao/quantization/fx/utils.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:979]: maybe we can put the logic to get observer into a helper function?

File: torch/ao/quantization/backend_config/qnnpack.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/backend_config/qnnpack.py:45]: I see, you mean require this one to be `per_tensor_symmetric` or `per_channel_symmetric`?

Comment (andrewor14) [reply torch/ao/quantization/backend_config/qnnpack.py:45]: Would we want to support affine qint8 weight with affine qint8 activations in the future? (right now we have symmetric qint8 weight with affine qint8 activations) @digantdesai

File: torch/ao/quantization/qconfig.py [hunk 3]
```diff
@@ -239,8 +240,8 @@ def get_default_qconfig(backend='fbgemm', version=0):
             qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=True),
                               weight=default_per_channel_weight_observer)
         elif backend == 'qnnpack':
-            qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=False),
-                              weight=default_weight_observer)
+            qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=False, eps=2 ** -12),
+                              weight=weight_observer_range_neg_127_to_127)
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig.py:244]: let's separate this change to another PR since this is bc-breaking

File: torch/ao/quantization/backend_config/qnnpack.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/backend_config/qnnpack.py]: > I see, you mean require this one to be per_tensor_symmetric or per_channel_symmetric?

yeah. this should be a separate PR though

File: torch/ao/quantization/backend_config/qnnpack.py [hunk 4]
```diff
@@ -11,17 +11,50 @@
...
+    input_dtype=xnnpack_act_qint8_with_constraints,
+    output_dtype=xnnpack_act_qint8_with_constraints,
+    weight_dtype=xnnpack_weight_qint8_with_constraints,
+    bias_dtype=torch.float,
+)
+
+xnnpack_default_op_qint8_dtype_config = DTypeConfig(
+    input_dtype=xnnpack_act_qint8_with_constraints,
+    output_dtype=xnnpack_act_qint8_with_constraints,
+)
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/qnnpack.py:55]: cc @digantdesai do we have the min_scale constraints on normal ops as well? or just linear and conv?

File: torch/ao/quantization/fx/utils.py [hunk 5]
```diff
@@ -954,6 +954,18 @@ def find_patterns(
...
+    activation_post_process: Union[ObserverBase, FakeQuantize],
+) -> ObserverBase:
+    """
+    If `activation_post_process` is an observer, return the observer.
+    If `activation_post_process` is a fake quantize, return the internal observer.
+    """
+    if isinstance(activation_post_process, ObserverBase):
+        return activation_post_process
+    else:
+        return activation_post_process.activation_post_process  # type: ignore[return-value]
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:967]: nit: assert that this is a FakeQuantizeBase?

File: torch/ao/quantization/fx/utils.py [hunk 6]
```diff
@@ -954,6 +954,18 @@ def find_patterns(
         for user in list(last_getitem.users.keys()):
             user.replace_input_with(last_getitem, new_input)
 
+def _get_observer_from_activation_post_process(
+    activation_post_process: Union[ObserverBase, FakeQuantize],
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:958]: FakeQuantize --> FakeQuantizeBase

File: torch/ao/quantization/fx/utils.py [hunk 7]
```diff
@@ -972,11 +984,12 @@ def _activation_post_process_satisfies_dtype_config_constraints(
             activation_post_process: Union[ObserverBase, FakeQuantize],
             dtype_with_constraints: DTypeWithConstraints,
             debug_string: str) -> bool:
-        app_quant_min = getattr(activation_post_process, "quant_min", None)
-        app_quant_max = getattr(activation_post_process, "quant_max", None)
+        observer = _get_observer_from_activation_post_process(activation_post_process)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:987]: I guess the type in L984 should also be FakeQuantizeBase as well

**Review by jerryzh168 — APPROVED** [review]
looks good, had some comments inline

File: torch/ao/quantization/backend_config/qnnpack.py [hunk 8]
```diff
@@ -79,16 +112,32 @@ def get_qnnpack_backend_config() -> BackendConfig:
     """
     Return the `BackendConfig` for PyTorch's native QNNPACK backend.
     """
-    conv_dtype_configs = [qnnpack_weighted_op_int8_dtype_config]
+    conv_dtype_configs = [
+        xnnpack_weighted_op_qint8_dtype_config,
+        qnnpack_weighted_op_int8_dtype_config,
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/qnnpack.py:117]: might make sense to rename int8 to something more specific like quint8

**Comment by vkuzo** [comment]
Just curious, what's the user story for this that we want?  Is it:
1. qnnpack is user facing, xnnpack may run some things behind the scenes but user shouldn't care
2. both qnnpack and xnnpack is user facing

Is it (1) or (2), or something else?  From previous conversations with @kimishpatel  and @digantdesai  I remember we were going for (1) - does that sound right?

If it is in fact (1), I think it would be good to name the user facing variables in terms of qnnpack and only mention xnnpack in the comments.  I think the way things named now is a bit confusing, something like `xnnpack_act_qint8_with_constraints` implies that `xnnpack` is user facing.  If we want to do (1), how about `qnnpack_act_qint8_with_constraints_compatible_with_xnnpack_ops`?

File: torch/ao/quantization/backend_config/qnnpack.py [hunk 9]
```diff
@@ -11,17 +11,50 @@
...
+
+# TODO: add additional restriction on qscheme to ensure it
+# is either per_tensor_symmetric or per_channel_symmetric
+
+xnnpack_act_qint8_with_constraints = DTypeWithConstraints(
+    dtype=torch.qint8,
+    scale_min_lower_bound=2 ** -12,
+)
+
+xnnpack_weight_qint8_with_constraints = DTypeWithConstraints(
```
Comment (kimishpatel) [inline torch/ao/quantization/backend_config/qnnpack.py:38]: I dont think we should pollute quantization codebase with backend information anymore than what is already there. So lets not introduce xnnpack here. Can we not just be explicit and name this `weight_qint8_with_symmetric_constraints` and if we want to use backend here maybe say `weight_qint8_with_qnnpack_symmetric_constraints`. Ideally I would like to say `weight_qint8_symmetric_min_scale_2_12` but that is mouthful and confusing.

**Comment by kimishpatel** [comment]
> 

Agree with @vkuzo. 1 is what we agreed. As much as possible quant codebase, including user facing api, should not expose xnnpack as yet another backend.

**Comment by andrewor14** [comment]
Sounds good. Just wanted to clarify that xnnpack_act_qint8_with_constraints is not actually user facing (not exposed through __all__), but I can rename it and limit the reference to xnnpack to comments.

**Comment by vkuzo** [comment]
> weight_qint8_symmetric_min_scale_2_12

IMO this is totally fine, the variable name should describe the contents.

**Comment by vkuzo** [comment]
> Just wanted to clarify that xnnpack_act_qint8_with_constraints is not actually user facing (not exposed through __all__), but I can rename it and limit the reference to xnnpack to comments.

If the story is "there is no user facing xnnpack backend", I think we shoulnd't have any Python variables named after xnnpack.  It should be "backend is qnnpack, and sometimes xnnpack ops are used" consistently throughout the codebase, for both users and developers.  I do think that's a bit confusing, but that's the story we agreed on so we should be consistent.

**Comment by andrewor14** [comment]
Thanks everyone for the feedback. I renamed these to `qnnpack_act_qint8_scale_min_2_neg_12` and `qnnpack_weight_qint8_neg_127_to_127_scale_min_2_neg_12`.
